signer/contentsignaturepki/contentsignature.go

```diff
@@ -79,82 +80,109 @@ func New(conf signer.Configuration) (s *ContentSigner, err error) {
+		return nil, fmt.Errorf("contentsignaturepki %q: invalid public key type for issuer, must be ecdsa", s.ID)
 	}
 	s.Mode = s.getModeFromCurve()
 
-	// the end-entity key is not stored in configuration but may already
-	// exist in an hsm, if present. Try to retrieve it, or make a new one.
-	var tx *database.Transaction
-	if s.db != nil {
-		tx, err = s.db.BeginEndEntityOperations()
-		if err != nil {
-			return nil, errors.Wrap(err, "contentsignaturepki: failed to begin end-entity db operations")
-		}
-	}
-	err = s.findAndSetEE(conf, tx)
+	err = s.initEE(conf)
 	if err != nil {
-		if err == database.ErrNoSuitableEEFound {
-			log.Printf("contentsignaturepki: making new end-entity for signer %q", s.ID)
-			// create a label and generate the key
-			s.eeLabel = fmt.Sprintf("%s-%s", s.ID, time.Now().UTC().Format("20060102150405"))
-			s.eePriv, s.eePub, err = conf.MakeKey(s.issuerPub, s.eeLabel)
+		return nil, errors.Wrapf(err, "contentsignaturepki %q: failed to initialize end-entity", s.ID)
+	}
+	return
+}
+
+// initEE configures an end-entity key and certificate that will be used
+// for signing. It will try to retrieve an existing one from db/hsm, and if
+// no suitable candidate can be found, a new one will be created.
+func (s *ContentSigner) initEE(conf signer.Configuration) error {
+	err := s.findAndSetEE(conf)
+	switch err {
+	case nil:
+		log.Printf("contentsignaturepki %q: reusing existing EE %q", s.ID, s.eeLabel)
+	case database.ErrNoSuitableEEFound:
+		// No suitable end-entity found, making a new chain
+		log.Printf("contentsignaturepki %q: making new end-entity", s.ID)
+		var tx *database.Transaction
+		if s.db != nil {
+			tx, err = s.db.BeginEndEntityOperations()
 			if err != nil {
-				err = errors.Wrap(err, "failed to generate key for end entity")
-				return
+				return errors.Wrapf(err, "contentsignaturepki %q: failed to begin db operations", s.ID)
 			}
-			// make the certificate and upload the chain
-			err = s.makeAndUploadChain()
+		}
+		// to prevent race conditions, we perform another set of the EE just in case
+		// someone else created it before we managed to obtain the lock
+		err = s.findAndSetEE(conf)
+		switch err {
+		case nil:
+			// alright we found a suitable EE this time to don't make one
+			goto releaseLock
+		case database.ErrNoSuitableEEFound:
+			// still nothing suitable, continue on
+			break
+		default:
+			// some other error popped up, exit
+			return err
+		}
+		// create a label and generate the key
+		s.eeLabel = fmt.Sprintf("%s-%s", s.ID, time.Now().UTC().Format("20060102150405"))
+		s.eePriv, s.eePub, err = conf.MakeKey(s.issuerPub, s.eeLabel)
+		if err != nil {
+			return errors.Wrapf(err, "contentsignaturepki %q: failed to generate end entity", s.ID)
+		}
+		// make the certificate and upload the chain
+		err = s.makeAndUploadChain()
+		if err != nil {
+			return errors.Wrapf(err, "contentsignaturepki %q: failed to make chain and x5u", s.ID)
+		}
```

Comment:
nit: logging the chain and upload location on success would be useful

Reply:
done!